### PR TITLE
[tensorflow/compiler/mlir/xla/ir/mlir_hlo_builder.cc] Add calls to `reserve()` before populating vectors

### PR DESCRIPTION
**Pull request by SamuelMarks**
https://github.com/tensorflow/tensorflow/pull/51739#issuecomment-940389562 told me to split the larger PR into one PR per file; thus this (thanks `bash`, `git` and `gh`!)

### CODE REVIEW

**Comment by gbaned** [comment]
@SamuelMarks Any update on this PR? Please. Thanks!

**Comment by SamuelMarks** [comment]
@gbaned I had this follow up question to @joker-eph https://github.com/tensorflow/tensorflow/pull/52473#discussion_r734204686

Also I'm not sold on the lack of `reserve`s for `high`, `min`, and `low`…

**Comment by joker-eph** [comment]
> @gbaned I had this follow up question to @joker-eph [#52473 (comment)](https://github.com/tensorflow/tensorflow/pull/52473#discussion_r734204686)
> 
> Also I'm not sold on the lack of `reserve`s for `high`, `min`, and `low`…

Sure: if you can collect statistics that shows SmallVector does not do the job I'll be happy to reconsider.

File: tensorflow/compiler/mlir/xla/ir/mlir_hlo_builder.cc [hunk 1]
```diff
@@ -561,10 +561,13 @@ StatusOr<XlaOp> MlirHloBuilder::PadInternal(
   TF_ASSIGN_OR_RETURN(
       mlir::Type result_type,
       ConvertShapeToType<mlir::RankedTensorType>(shape, builder_));
-  std::vector<int64_t> low;
-  std::vector<int64_t> high;
-  std::vector<int64_t> internal;
-  for (auto& dimension : padding_config.dimensions()) {
+  auto dimensions = padding_config.dimensions();
```
[the remark this answers is not in the document]
Comment (joker-eph) [reply tensorflow/compiler/mlir/xla/ir/mlir_hlo_builder.cc:564]: I'm not sure if this is right, if it compiles it is likely correct. Not sure I'd write this like that though, it's pretty verbose.

[the remark this answers is not in the document]
Comment (SamuelMarks) [reply tensorflow/compiler/mlir/xla/ir/mlir_hlo_builder.cc:564]: Yep but it matches your request to:
> Spell out the type here please.

File: tensorflow/compiler/mlir/xla/ir/mlir_hlo_builder.cc [hunk 2]
```diff
@@ -561,10 +561,13 @@ StatusOr<XlaOp> MlirHloBuilder::PadInternal(
...
+  auto dimensions_size = dimensions.size();
+  llvm::SmallVector<int64_t> low, high, internal;
+  low.reserve(dimensions_size);
+  high.reserve(dimensions_size);
+  internal.reserve(dimensions_size);
+  for (auto& dimension : dimensions) {
```
[the remark this answers is not in the document]
Comment (SamuelMarks) [reply tensorflow/compiler/mlir/xla/ir/mlir_hlo_builder.cc]: [thread has continued outside source comment… also master now has `llvm::SmallVector<int64_t> low, high, internal;` merged]

**Comment by joker-eph** [comment]
Closing now, please re-open if you gather some data to justify turning the one line into 6.

**Comment by SamuelMarks** [comment]
@joker-eph It seems like a hard piece of code to benchmark, I would need to know the various values of `padding_config.dimensions().size()` for different workloads in order to build a reasonable only-stl-and-llvm dependent test-case for this. So, unless I'm missing something, your call to benchmark was setting me up to fail

**Comment by joker-eph** [comment]
You don't need to run benchmarks, as in get *performance* numbers. I'm asking statistics in terms of "how many times does this code path triggers a grow of the vector".

Basically it could be done  by adding after the loop some code that looks like:

```
  {
    string temp_file;
     if (low.isSmall()) temp_file = tensorflow::io::GetTempFilename("low_nogrow.txt"));
     else temp_file = tensorflow::io::GetTempFilename("low_grow.txt"));
     TF_CHECK_OK(WriteStringToFile(tf::Env::Default(), temp_file, absl::StrCat("Low size is ", low.size()));
  }
```

Then run TensorFlow tests suite with `--test_env=TEST_TMPDIR=$HOME/grow_results` (you need to create the directory ahead of time)

Which should create a bunch of files in this directory which makes it for trivial statistics.

**Pull request closed without merging**